core/chains/evm/gas/universal_estimator.go

```diff
@@ -0,0 +1,412 @@
+
+	gasPrice, err := u.client.SuggestGasPrice(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch gas price: %s", err)
+	}
+
+	promUniversalEstimatorGasPrice.WithLabelValues(u.chainID.String()).Set(float64(gasPrice.Int64()))
+
+	gasPriceWei := (*assets.Wei)(gasPrice)
+
+	u.logger.Debugf("fetched new gas price: %v", gasPriceWei)
+
+	u.gasPriceMu.Lock()
+	defer u.gasPriceMu.Unlock()
+	u.gasPrice = gasPriceWei
+	return u.gasPrice, nil
+}
+
+func (u *UniversalEstimator) getGasPrice() (*assets.Wei, error) {
+	u.gasPriceMu.RLock()
+	defer u.gasPriceMu.RUnlock()
+	if u.gasPrice == nil {
+		return u.gasPrice, fmt.Errorf("gas price not set")
+	}
+	return u.gasPrice, nil
+}
+
+// GetDynamicFee will fetch the cached dynamic prices.
+func (u *UniversalEstimator) GetDynamicFee(ctx context.Context, maxPrice *assets.Wei) (fee DynamicFee, err error) {
+	if fee, err = u.getDynamicPrice(); err != nil {
+		return
+	}
+
+	if fee.FeeCap == nil || fee.TipCap == nil {
+		return fee, fmt.Errorf("dynamic price not set")
+	}
+	if fee.FeeCap.Cmp(maxPrice) > 0 {
+		u.logger.Warnf("estimated maxFeePerGas: %v is greater than the maximum price configured: %v, returning the maximum price instead.",
+			fee.FeeCap, maxPrice)
+		fee.FeeCap = maxPrice
+		if fee.TipCap.Cmp(maxPrice) > 0 {
+			u.logger.Warnf("estimated maxPriorityFeePerGas: %v is greater than the maximum price configured: %v, returning the maximum price instead.",
+				fee.TipCap, maxPrice)
+			fee.TipCap = maxPrice
+		}
+	}
+
+	return
+}
+
+// FetchDynamicPrice uses eth_feeHistory to fetch the basFee of the latest block and the Nth maxPriorityFeePerGas percentiles
+// of the past X blocks. It also fetches the highest Zth maxPriorityFeePerGas percentile of the past X blocks. Z is configurable
+// and it represents the highest percentile we're willing to pay.
+// A buffer is added on top of the latest basFee to catch fluctuations in the next blocks. On Ethereum the increase is baseFee*1.125 per block.
+func (u *UniversalEstimator) FetchDynamicPrice() (fee DynamicFee, err error) {
+	ctx, cancel := u.stopCh.CtxCancel(evmclient.ContextWithDefaultTimeout())
+	defer cancel()
+
+	if u.config.BlockHistorySize == 0 {
+		return fee, fmt.Errorf("BlockHistorySize cannot be 0")
+	}
+	// RewardPercentile will be used for maxPriorityFeePerGas estimations and connectivityPercentile to set the highest threshold for bumping.
+	feeHistory, err := u.client.FeeHistory(ctx, u.config.BlockHistorySize, []float64{u.config.RewardPercentile, ConnectivityPercentile})
+	if err != nil {
+		return fee, fmt.Errorf("failed to fetch dynamic prices: %s", err)
+	}
+
+	// Latest base fee
+	baseFee := (*assets.Wei)(feeHistory.BaseFee[len(feeHistory.BaseFee)-1])
+	latestBlock := big.NewInt(0).Add(feeHistory.OldestBlock, big.NewInt(int64(u.config.BlockHistorySize)))
+
+	priorityFee := big.NewInt(0)
+	priorityFeeThreshold := big.NewInt(0)
+	for _, fee := range feeHistory.Reward {
+		priorityFee = priorityFee.Add(priorityFee, fee[0])
```

Comment:
Should we consider filtering out blocks where 0 priority fee is returned? I could see the argument to leave them in since an empty block technically would have required 0 priority fee to be included in. But I could also see a scenario where we get a long streak of empty blocks so we set our cache to 0 then fall behind the market priority fee. I've only noticed streaks of empty blocks on Amoy and also considering a priority fee doesn't make too big of a difference in inclusion time, I'm leaning towards just leaving it the way it is.

Reply:
Good call out. This will be in conjunction with the BlockHistoryRange, so for chains that don't have many transactions we should increase the range. But as you mentioned, if a chain consistently doesn't include transactions in blocks, it means whatever `priorityFee` we use, our tx should be included because there is no traffic. Also for certain chains that don't rely on `priorityFee` at all we have a way to skip it entirely.